### PR TITLE
add nodepool support for max_pods_per_node

### LINKED ISSUE #902: Allow for node pool specific max_pods_per_node
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The setting max_pods_per_node is currently a global variable and should be scoped down to each node pool.

### New or Affected Resource(s)
[VPC-native networking
](https://github.com/oracle-terraform-modules/terraform-oci-oke/blob/main/modules/workers/nodepools.tf#L60)

### Potential Terraform Configuration

```hcl
# bad
worker_pools = {
    azdo-build-agents-cpu= {
      allow_autoscaler = true,
      max_pods_per_node = var.max_pods_per_node
      cni_type         = "npn" 
    },
    azdo-build-agents-ram = {
      allow_autoscaler = true,
      max_pods_per_node =  var.max_pods_per_node
      cni_type         = "npn" 
    }
  }
  # ok
  worker_pools = {
    azdo-build-agents-cpu= {
      allow_autoscaler = true,
      max_pods_per_node = 4
      cni_type         = "npn"
    },
    azdo-build-agents-ram = {
      allow_autoscaler = true,
      max_pods_per_node = 16
      cni_type         = "npn" 
    }
  }
```

### PR DESCRIPTION
**Pull request by robo-cap**
fixes #902 .

### CODE REVIEW

File: modules/workers/nodepools.tf [hunk 1]
```diff
@@ -57,7 +57,7 @@ resource "oci_containerengine_node_pool" "workers" {
       for_each = var.cni_type == "npn" ? [1] : []
       content { # VCN-Native requires max pods/node, nsg ids, subnet ids
         cni_type          = "OCI_VCN_IP_NATIVE"
-        max_pods_per_node = min(max(var.max_pods_per_node, 1), 110)
+        max_pods_per_node = each.value.max_pods_per_node
```
Comment (hyder) [inline modules/workers/nodepools.tf:60]: If the node pool does not specify this value, it should take the global default.

File: modules/workers/nodepools.tf [hunk 2]
```diff
@@ -57,7 +57,7 @@ resource "oci_containerengine_node_pool" "workers" {
       for_each = var.cni_type == "npn" ? [1] : []
       content { # VCN-Native requires max pods/node, nsg ids, subnet ids
         cni_type          = "OCI_VCN_IP_NATIVE"
```
Comment (hyder) [inline modules/workers/nodepools.tf:59]: Shouldn't you specify this for flannel as well?

**Review by hyder — CHANGES_REQUESTED** [review]
1. Question about whether this should be added for flannel too
2. Global default should be used when the node pool does not specify the value

File: modules/workers/nodepools.tf (hunk 1 above)
Comment (robo-cap) [reply modules/workers/nodepools.tf:60]: Yes, the global value is defined in the `worker_pool_defaults`, [here](https://github.com/oracle-terraform-modules/terraform-oci-oke/pull/904/files#diff-54c5df4ead89ff4bd0fb96b9fba84ff8206c61a5b27be4c7f8f2d6124e810a3fR42).

File: modules/workers/nodepools.tf (hunk 2 above)
Comment (robo-cap) [reply modules/workers/nodepools.tf:59]: The attribute is applicable only for npn: https://registry.terraform.io/providers/oracle/oci/latest/docs/resources/containerengine_node_pool#max_pods_per_node

File: modules/workers/nodepools.tf (hunk 1 above)
Comment (hyder) [reply modules/workers/nodepools.tf:60]: I'm not seeing where you are taking the global default if the value in the nodepool is not specified. 

You network options look like this:

```
        max_pods_per_node = each.value.max_pods_per_node
```

You should have a coalesce here: 

```
        max_pods_per_node = coalesce(each.value.max_pods_per_node, local.max_pods_per_node)
```

**Review by hyder — CHANGES_REQUESTED** [review]
please check again

**Review by hyder — APPROVED** [review]
approved

**Pull request merged**